src/mono/wasm/runtime/assets.ts

```diff
@@ -181,52 +245,94 @@ async function start_asset_download_sources(asset: AssetEntry): Promise<AssetEnt
                 hash: asset.hash,
                 behavior: asset.behavior
             });
-            asset.pending = loadingResource;
+            asset.pendingDownloadInternal = loadingResource;
             response = await loadingResource.response;
             if (!response.ok) {
                 continue;// next source
             }
-            ++downloded_assets_count;
-            return asset;
+            ++actual_downloaded_assets_count;
+            return response;
         }
         catch (err) {
             continue; //next source
         }
     }
-    throw response;
+    const isOkToFail = asset.isOptional || (asset.name.match(/\.pdb$/) && runtimeHelpers.config.ignorePdbLoadErrors);
+    mono_assert(response, () => `Response undefined ${asset.name}`);
+    if (!isOkToFail) {
+        const err: any = new Error(`MONO_WASM: download '${response.url}' for ${asset.name} failed ${response.status} ${response.statusText}`);
+        err.status = response.status;
+        throw err;
+    } else {
+        Module.print(`MONO_WASM: download '${response.url}' for ${asset.name} failed ${response.status} ${response.statusText}`);
+        if (!skipDownloadsByAssetTypes[asset.behavior]) {
+            expected_downloaded_assets_count--;
+        }
+        if (skipInstantiateByAssetTypes[asset.behavior]) {
+            expected_instantiated_assets_count--;
+        }
```

Comment:
How can we get here? It's already guarded in the `mono_download_assets`, right?

Reply:
it's about optional downloads, I moved it now.